### PR TITLE
Bluetooth: att: Fix EATT channel security requirements

### PR DESCRIPTION
**Pull request by MariuszSkamra**
Core Vol 3, Part G, Section 5.3.2 Channel Requirements states that
"The channel shall be encrypted". It does not mention any additional
security requirements that can be specified bt higher layer profiles.
This enables link encryption requirement for EATT channel.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

### CODE REVIEW

**Comment by MariuszSkamra** [comment]
> The change seems OK, but I was wondering: Why do we not just ensure that the ACL is encrypted before allowing EATT channels to be established? If we cannot use EATT while not encrypted, then there's not really any reason to even establish them. 

You have right. The fix would be even cleaner, as there the check won't be evaluated on every PDU. Thanks!

**Comment by Thalley** [comment]
> > The change seems OK, but I was wondering: Why do we not just ensure that the ACL is encrypted before allowing EATT channels to be established? If we cannot use EATT while not encrypted, then there's not really any reason to even establish them.
> 
> You have right. The fix would be even cleaner, as there the check won't be evaluated on every PDU. Thanks!

We may want to check if this causes some IOP issues: Does the core spec say anything about encryption level before establishing EATT channels? We may experience some IOP issues if we change it to that. 

The auto-EATT connection we have in Zephyr would also need to be changed from when it's connected (done in `bt_att_accept` now), to when the ACL has been encrypted. 

@hermabe Any input here?

**Review by jhedberg — DISMISSED** [review]

**Review by Thalley — DISMISSED** [review]
Assuming that this won't cause IOP issues, LGTM to me

File: tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/gatt_server_test.c [hunk 1]
```diff
@@ -103,12 +122,7 @@ static void test_main_common(bool connect_eatt)
 	WAIT_FOR_FLAG(flag_is_connected);
 
 	if (connect_eatt) {
-		err = bt_eatt_connect(g_conn, CONFIG_BT_EATT_MAX);
-		if (err) {
-			FAIL("Failed to connect EATT channels (err %d)\n", err);
-
-			return;
-		}
+		WAIT_FOR_FLAG(flag_eatt_connect_success);
```
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/gatt_server_test.c:125]: It does seem like a WAIT_FOR_FLAG(flag_is_encrypted) before calling `bt_eatt_connect` would be better than calling `bt_eatt_connect` in the callback itself (fewer lines changed, more similar to your change in the other test, and the `security_changed` does not take `connect_eatt` into account

**Review by Thalley — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/gatt_server_test.c (hunk 1 above)
Comment (MariuszSkamra) [reply tests/bluetooth/bsim_bt/bsim_test_gatt_caching/src/gatt_server_test.c]: indeed

**Comment by MariuszSkamra** [comment]
> The spec is really vague on this requirement. It only says "the channel shall be encrypted". I haven't seen any special requirement that the link shall be encrypted before the channels are connected. It may be ok to connect, but not send over EATT channels without encryption (the current implementation). In that case, we may want to allow incoming requests/etc. without encryption for IOP reasons.

- I agree, it is vague. However, if you allow to connect and then reject requests with ATT Insufficient Encryption, then you are in trouble because you are against the specification that does not allow such error when finding attributes:

Core 5.3 | Vol, Part F, 3.4.9 Attribute PDU response summary
image

The Core is strict here:
Core 5.3 | Vol 3, Part F, 3.4.3.3 ATT_FIND_BY_TYPE_VALUE_REQ
image

ATT_FIND_BY_TYPE_VALUE_REQ is PDU used by GATT Discover Primary Service by Service UUID procedure.

- Moreover, GATT Discover All Primary Services procedure uses ATT_READ_BY_GROUP_TYPE_REQ PDU, which can return ATT Insufficient Encryption error, but if the attribute permissions require encryption to be enabled.

image

but GATT Service Declaration *shall* not require authentication as per Core 5.3 | Vol 3, Part G, 3.1 SERVICE DEFINITION.
image

> 
> Requiring the link to be encrypted before the channels are established is stricter, but should be within the spec. This is also cleaner as the peer gets an auth error code in response to the connection request.
> 

The GATT requirements I mentioned above imply the link to be encrypted when EATT channels are established. I will create an errata to Core specification with clarification request and paste link here

> From some brief testing with my phone (Android 12) some time ago, the phone did not check the encryption requirement and allowed notifications to be received over EATT without encryption.

Yes, this will be fixed soon, as my colleague is working on it now.

**Comment by MariuszSkamra** [comment]
Errata: https://www.bluetooth.org/errata/errata_view.cfm?errata_id=19172

**Review by Thalley — APPROVED** [review]

**Comment by sjanc** [comment]
#AutoPTS run zephyr GATT

**Comment by codecoup-tester** [comment]
Scheduled PR #https://github.com/zephyrproject-rtos/zephyr/pull/46982#issuecomment-1171266517 after 16:09:00.

**Comment by codecoup-tester** [comment]
AutoPTS Bot results:
Failed tests:
GATT    GATT/CL/GAR/BI-42-C             INCONC
GATT    GATT/CL/GAR/BV-10-C             FAIL
GATT    GATT/CL/GAR/BV-11-C             INCONC
GATT    GATT/CL/GPM/BV-11-C             FAIL
GATT    GATT/SR/GAN/BV-02-C             INCONC
GATT    GATT/SR/GAN/BV-02-C_LT2         FAIL

**Comment by MariuszSkamra** [comment]
> AutoPTS Bot results: Failed tests: GATT GATT/CL/GAR/BI-42-C INCONC GATT GATT/CL/GAR/BV-10-C FAIL GATT GATT/CL/GAR/BV-11-C INCONC GATT GATT/CL/GPM/BV-11-C FAIL GATT GATT/SR/GAN/BV-02-C INCONC GATT GATT/SR/GAN/BV-02-C_LT2 FAIL

@sjanc Those failed in Bluetooth test session - WW26.3 as well

**Comment by Thalley** [comment]
> @hermabe I proposed to change "The channel shall be encrypted." into "The channel shall be created and used on encrypted link" https://www.bluetooth.org/errata/errata_view.cfm?errata_id=19172 That's the response I got: "However, I don't think we need to say that the connection must be encrypted before channel creation; if that's the only way of getting an encrypted channel, then it's implicit and we don't (or shouldn't) duplicate requirements in specifications."

Hmm, that's a shame. 

w.r.t. IOP issues, then this shouldn't give any _real_ issues: Worst case scenario is that the EATT channels are not established, and there isn't any usecase (yet) that is "EATT only". 
In the case that we (Zephyr) reject at EATT request due to lacking encryption, then we'll (Zephyr) simply just attempt to create the EATT channels once the link is encrypted, so in the end the channels will probably be created.

**Review by carlescufi — APPROVED** [review]

**Review by sjanc — APPROVED** [review]

**Review by hermabe — APPROVED** [review]
LGTM. This approach seems cleaner than connecting EATT channels that cannot be used.

**Pull request merged**